File: src/main/java/gargoyle/ct/ui/impl/control/CTLocalizableMenuItem.java
package gargoyle.ct.ui.impl.control;

import gargoyle.ct.messages.LocaleProvider;
import org.jetbrains.annotations.NotNull;

import javax.swing.*;

public final class CTLocalizableMenuItem extends JMenuItem {
    private static final long serialVersionUID = -3628913101849581469L;

    public CTLocalizableMenuItem(LocaleProvider provider, @NotNull CTLocalizableAction action) {
        super(action);
        provider.locale().addPropertyChangeListener(event -> update(action));
        action.init(this);
    }

    private void update(CTLocalizableAction action) {
        setText(action.getLocalizedText());
        setToolTipText(action.getToolTipLocalizedText());
    }
}

File: src/main/kotlin/gargoyle/ct/ui/impl/control/CTLocalizableMenuItem.kt
package gargoyle.ct.ui.impl.control

import gargoyle.ct.messages.LocaleProvider
import java.io.Serial
import javax.swing.JMenuItem

class CTLocalizableMenuItem(provider: LocaleProvider, action: CTLocalizableAction) : JMenuItem(action) {
    init {
        provider.locale()
            .addPropertyChangeListener { update(action) }
        action.init(this)
    }

    private fun update(action: CTLocalizableAction) {
        text = action.localizedText
        toolTipText = action.toolTipLocalizedText
    }

    companion object {
        @Serial
        private val serialVersionUID = -3628913101849581469L
    }
}
